group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ClientAssignor.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.coordinator.group.consumer;
+
+import org.apache.kafka.coordinator.group.generated.ConsumerGroupMemberMetadataValue;
+
+import java.nio.ByteBuffer;
+import java.util.Objects;
+
+/**
+ * An immutable representation of a client side assignor within a consumer group member.
+ */
+public class ClientAssignor {
+    /**
+     * The name of the assignor.
+     */
+    private final String name;
+
+    /**
+     * The reason reported by the assignor.
+     */
+    private final byte reason;
+
+    /**
+     * The minimum metadata version supported by the assignor.
+     */
+    private final short minimumVersion;
+
+    /**
+     * The maximum metadata version supported by the assignor.
+     */
+    private final short maximumVersion;
+
+    /**
+     * The versioned metadata.
+     */
+    private final VersionedMetadata metadata;
```

Comment:
what's an example of a metadata stored inside the client assignor?

Reply:
This is not defined and left to the implementation of the assignor. The foreseen schema for Streams are defined [here](https://cwiki.apache.org/confluence/display/KAFKA/KIP-848%3A+The+Next+Generation+of+the+Consumer+Rebalance+Protocol#KIP848:TheNextGenerationoftheConsumerRebalanceProtocol-StreamsMemberMetadataandAssignmentMetadata).

Reply:
to confirm, client assignor implementations will be under the `consumer` package?

Reply:
correct but metadata flows via the protocol/coordinator.

Reply:
I am not sure to understand what you would like to see. `ClientAssignor` is all about the client assignor so everything in this class is about this.

Reply:
```
    /**
     * The error assigned to the member.
     */
    private final byte error;
```

you mean more than this?

Reply:
sorry to have hijacked this thread, but something like
```
    /**
     * The error assigned to the member. Depends on the client assignor 
     * implementation and commonly used for Streams
     */
    private final byte error;
```

if you think what we have is sufficient, i am fine with it. just wanted to mention that i didn't fully understand how it would be used